File: c++/076.cpp
class Solution {
/*
滑动窗口的标准解题方式：
第一步使得窗口包含子串，
第二步使得窗口再次不包含子串
第三部重复上述一二步骤，记录找的最小的一个子串。

[c++实现](https://leetcode.cn/problems/minimum-window-substring/solution/leetcode-76-zui-xiao-fu-gai-zi-chuan-cja-lmqz/#:~:text=%E6%98%AF%20O(n)%E3%80%82-,2%E3%80%81c%2B%2B%E4%BB%A3%E7%A0%81,-class%20Solution%20%7B)
关键：
* 用两个哈希表维护：字符串t的字符 ht，窗口内的字符 hs
* 窗口为  [ j, i ]，（左右均闭）
* 变量 cnt 维护窗口中满足要求的字符总数（不超过 t）
* 收缩过程：i 一直向右扩展；当 j 指向的字符在窗口中的数量 > t 中数量时，j 向右收缩，同时窗口中字符 j 数量 -1；当窗口中满足要求的字符数量 cnt==len(t)，记录下当前子串，并判断是否为当前最小子串
*/

public:
    string minWindow(string s, string t) {
        unordered_map<char, int> hs, ht;
        for (auto c: t) ht[c] ++ ;
        string res;
        int cnt = 0;
        for (int i = 0, j = 0; i < s.size(); i ++ ) {
            hs[s[i]] ++ ;
            if (hs[s[i]] <= ht[s[i]]) cnt ++ ;

            while (hs[s[j]] > ht[s[j]]) hs[s[j ++ ]] -- ;
            if (cnt == t.size()) {
                if (res.empty() || i - j + 1 < res.size())
                    res = s.substr(j, i - j + 1);
            }
        }
        return res;
    }
};

File: go/076.go
package main

/**
最小覆盖子串
字符串 s t 从 s 的子串中找出包含 t 中所有字符的最小字串 一对一
全英文
暴力：每个字串进行比较
滑动窗口：维护窗口  一般用于匹配
i j 窗口左右边界
窗口在 s 上滑动，左边固定，j 移动直到窗口含有 t 的所有字符，
尝试移动 i 使得窗口依旧符合要求
判断是否符合要求 使用哈希表记录，每个字符的个数
技巧：存入 t 的字符，出现一次 -1，窗口每次滑动时 +1
这样没有匹配时该字符不是匹配字符是 >0 是的话 <= 0
这样记录每次匹配的次数 如果 cnt == len(t) 即代表窗口包含t所有字符
按照思路 该移动 i 了 如果不是匹配字符即可移动 即 >0
最后取最小字符串即可
*/
func minWindow(s string, t string) string {
	n, m := len(s), len(t)
	hash := [128]int{}
	for _, v := range t {
		hash[v]--
	}
	res := ""
	for i, j, cnt := 0, 0, 0; j < n; j++ {
		hash[s[j]]++
		if hash[s[j]] <= 0 {
			cnt++
		}
		if cnt == m {
			for hash[s[i]] > 0 {
				hash[s[i]]--
				i++
			}
			if res == "" || j-i+1 < len(res) {
				res = s[i : j+1]
			}
		}
	}
	return res
}
